backend/server/routers/programs.py

```diff
@@ -240,17 +239,87 @@ def get_structure(
                             "ARTS1031",
                             "ARTS1032",
                             "ARTS1060",
-                           
                         ]
-                    
                     }
                 }
             },
         },
     },
 )
-
 def getGenEds(
     programCode: str):
     all_geneds = data_helpers.read_data("data/scrapers/genedPureRaw.json")
     return {"courses" : all_geneds[programCode]}
+
+@router.get("/graph")
+def courses_graph():
+    """
+    Constructs a structure for the frontend to use for the graphical
+    selector.
+    Currently returns the ingoing and outgoing edges for each course
+    that is a part of the courselist
+    """
+    return {
+        "message": "This endpoint is not implemented yet",
+    }
+
+###############################################################
+#                       End of Routes                         #
+###############################################################
+
+def course_list_from_structure(structure: Dict) -> List[str]:
+    """
+        Given a formed structure, return the list of courses
+        in that structure
+    """
+    courses = []
+    def __recursive_course_search(structure: Dict) -> List[str]:
+        """ Recursively search for courses in a structure """
+        print(structure.keys())
+        if not isinstance(structure, (list, dict)):
+            return
+        # For a list, recurse on all its object
+        if not isinstance(structure, dict):
+            return
+        for k, v in structure.items():
+            with suppress(KeyError):
+                if not isinstance(v, dict) or "rule" in v["type"]:
```

Comment:
Could we also exclude gened courses (if it doesn't already)?

Reply:
the constructor for the `structure` doesn't attach geneds so this route won't return geneds

Reply:
> Could we also exclude gened courses (if it doesn't already)?

the constructor for the structure doesn't attach geneds so this route won't return geneds

If in a future use-case, geneds need to be excluded before we get a course-list, imo its preferable to have a function to remove the geneds before they get sent to this function